### ISSUE TITLE
Should we use double precision for all calculations?

### ISSUE THREAD

**Issue by Peter9192** [ESMValGroup/ESMValCore#943]
In #685 (followed up in #940) we noticed coincidentily that calculations on single precision (`float32`, i.e. the default dtype for all cmor data) may cause inaccurate results. This is known and documented behaviour in numpy, e.g. [here](https://numpy.org/doc/1.19/reference/generated/numpy.mean.html):

> Note that for floating-point input, the mean is computed using the same precision the input has. Depending on the input data, this can cause the results to be inaccurate, especially for float32 (see example below). Specifying a higher-precision accumulator using the dtype keyword can alleviate this issue.

The question thus arises whether we should explicitly enforce a higher precision before doing any calculations on the data. This could lead to more accurate results, but it might also increase memory use.

**Comment by nielsdrost** [comment]
Thanks for digging up this issue @Peter9192!

@valeriupredoi did you ever use any double precision for calculations on purpose?

@bjlittle is there any consensus in Iris on if double precision is needed for the type of calculations we do?

**Comment by valeriupredoi** [comment]
> @valeriupredoi did you ever use any double precision for calculations on purpose?

No, I looked into this a long time ago and with the aim of reducing the memory load, I remember I looked at moving to a `float16` actually, nothing towards going up in precision but rather going as low as it would be sane for results.

I'd be curious to know what sort of results need `float64` precisions, I think that's overkill, but if there are any we should defo impose the required precision then. But a used case would be nice to look at first :+1:

**Comment by jvegasbsc** [comment]
Just to add a bit more confusion, people in modelling are thinking in reducing the precision because most of the time it is not needed. This is a paper about this from a BSC colleague https://gmd.copernicus.org/articles/12/3135/2019/

By the way, I'm against always using `float64`, but it may be a good idea to have a preprocessor function that allow us to change data `dtype` so we can experiment a bit.

**Comment by bouweandela** [comment]
We could add the datatype as an optional arument to the [`load`](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.load) preprocessor function, so people can specify the option in the recipe if they want to. All other preprocessor functions could then take the cubes through the computations without changing the data type and we could (optionally) convert back to float32 in the [`save`](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.save) function.
